File: web/src/lib/parsers/bradesco/pdfExtrato.ts
// Parser do extrato da conta corrente do Bradesco em PDF, no navegador
// (pdfjs-dist, via ../pdfText). Mesma lógica de
// parsers/bradesco/pdf_extrato.py (desktop) — ver ali para detalhes do
// formato observado (bloco de 3 linhas por lançamento, direção inferida do
// prefixo REM:/DES: ou das palavras do rótulo do tipo).
import type { BankParser, NormalizedRow } from '../types'
import { guessCategory, looksLikeInvestment } from '../base'
import { cleanDescription, parseBrlAmount, stripAccents } from '../common'
import { extractPdfText } from '../pdfText'
import { guessPaymentMethod } from './common'

const NOISE_PREFIXES = ['Bradesco Celular', 'Data:', 'Nome:', 'Extrato de:']
const TOTAL_RE = /^Total\s+[\d.,]+/
const COD_LANC_RE = /^(?:(\d{2}\/\d{2}\/\d{4})\s+)?COD\.?\s*LANC\.?/i
const AMOUNT_LINE_RE = /^(?:(\d{2}\/\d{2}\/\d{4})\s+)?(\S+)\s+([\d.,]+)\s+([\d.,]+)\s*$/
const KNOWN_TYPE_LABELS = [
  'PIX RECEBIDO', 'PIX ENVIADO', 'PIX QR CODE',
  'COMPRA ELO DEBITO', 'PAGTO ELETRON', 'TED', 'DOC', 'TRANSFERENCIA',
]

function isNoise(ln: string): boolean {
  const upper = stripAccents(ln).toUpperCase()
  if (NOISE_PREFIXES.some((p) => ln.startsWith(p))) return true
  if (upper.startsWith('DATA HISTORICO')) return true
  if (TOTAL_RE.test(ln)) return true
  return false
}

function isKnownTypeLabel(ln: string): boolean {
  const upper = stripAccents(ln).toUpperCase()
  return KNOWN_TYPE_LABELS.some((t) => upper.startsWith(t))
}

function parseDateBr(s: string): string {
  const [dd, mm, yyyy] = s.split('/')
  return `${yyyy}-${mm}-${dd}`
}

function inferDirection(tipo: string, detail: string): 'entrada' | 'saida' {
  const detailUpper = stripAccents(detail).toUpperCase()
  if (detailUpper.startsWith('REM:')) return 'entrada'
  if (detailUpper.startsWith('DES:')) return 'saida'
  const tipoUpper = stripAccents(tipo).toUpperCase()
  return tipoUpper.includes('RECEBIDO') ? 'entrada' : 'saida'
}

export function parseText(text: string): NormalizedRow[] {
  const rawLines = text.split(/\r?\n/).map((ln) => ln.trim()).filter(Boolean)
  const lines = rawLines.filter((ln) => !isNoise(ln))

  const rows: NormalizedRow[] = []
  let currentDate: string | null = null
  let tipoBuffer: string[] = []

  let i = 0
  const n = lines.length
  while (i < n) {
    const ln = lines[i]

    const codM = COD_LANC_RE.exec(ln)
    if (codM) {
      if (codM[1]) currentDate = parseDateBr(codM[1])
      tipoBuffer = []
      i += 1
      continue
    }

    const amtM = AMOUNT_LINE_RE.exec(ln)
    if (amtM) {
      if (amtM[1]) currentDate = parseDateBr(amtM[1])
      const tipo = cleanDescription(tipoBuffer.join(' '))
      tipoBuffer = []
      i += 1

      let detail = ''
      if (i < n) {
        const nxt = lines[i]
        if (!AMOUNT_LINE_RE.test(nxt) && !COD_LANC_RE.test(nxt) && !isKnownTypeLabel(nxt)) {
          detail = nxt
          i += 1
        }
      }

      if (tipo && currentDate !== null) {
        const amount = parseBrlAmount(amtM[3])
        const direction = inferDirection(tipo, detail)
        const desc = cleanDescription(detail) || tipo
        const isInv = looksLikeInvestment(`${tipo} ${detail}`)
        rows.push({
          date: currentDate,
          description: desc,
          amount: Math.abs(amount),
          direction,
          suggestedCategory: isInv ? 'Investimentos' : guessCategory(desc),
          suggestedPaymentMethod: guessPaymentMethod(`${tipo} ${detail}`),
          isInvestmentLike: isInv,
          isCreditCardCharge: false,
          raw: `${tipo} | ${ln} | ${detail}`.trim(),
        })
      }
      continue
    }

    tipoBuffer.push(ln)
    i += 1
  }

  return rows
}

export class BradescoPdfExtratoParser implements BankParser {
  bankId = 'bradesco'
  formatId = 'pdf_extrato'

  sniff(_bytes: ArrayBuffer, filename: string): boolean {
    // A checagem real (texto contém "Bradesco") só é possível de forma
    // assíncrona (extração de PDF) — aqui só filtra pela extensão;
    // registry.ts's detectAndParse() descarta o match se parse() vier vazio.
    return filename.toLowerCase().endsWith('.pdf')
  }

  async parse(bytes: ArrayBuffer): Promise<NormalizedRow[]> {
    const text = await extractPdfText(bytes)
    const upper = stripAccents(text).toUpperCase()
    if (!upper.includes('BRADESCO')) return []
    return parseText(text)
  }
}

File: parsers/bradesco/pdf_extrato.py
"""Parser do extrato da conta corrente do Bradesco em PDF.

Formato observado (amostra real, 2026-09): cada lançamento normal vira
um bloco de 3 linhas, em ordem fixa, na extração de texto do
pdfplumber:

    PIX RECEBIDO                              <- rótulo do tipo, sozinho
    09/09/2026 1715281 23,70 9.796,56          <- [data opcional] docto valor saldo
    REM: FULANO DA SILVA 07/09                 <- detalhe (REM:/DES: ou nome do estabelecimento)

A data só aparece na primeira transação de cada dia (linhas seguintes
do mesmo dia omitem). Nenhuma posição de coluna sobrevive à extração de
texto — não dá pra saber se um valor era crédito ou débito pela
posição; a direção é inferida do prefixo da linha de detalhe (REM: =
entrada, DES: = saída) com fallback nas palavras do rótulo do tipo
(RECEBIDO/ENVIADO/COMPRA/PAGTO). Linhas "COD. LANC. N" são marcador de
saldo, vêm inteiras numa linha só (sem bloco de 3 linhas) — são
puladas.
"""
import io
import re
from datetime import datetime
from typing import List, Optional

from parsers.base import (
    NormalizedRow, clean_description, guess_category, looks_like_investment,
    parse_brl_amount, strip_accents,
)
from parsers.bradesco.common import guess_payment_method

_NOISE_PREFIXES = ("Bradesco Celular", "Data:", "Nome:", "Extrato de:", "Data Historico")
_TOTAL_RE       = re.compile(r"^Total\s+[\d.,]+")
_COD_LANC_RE    = re.compile(r"^(?:(\d{2}/\d{2}/\d{4})\s+)?COD\.?\s*LANC\.?", re.IGNORECASE)
_AMOUNT_LINE_RE = re.compile(r"^(?:(\d{2}/\d{2}/\d{4})\s+)?(\S+)\s+([\d.,]+)\s+([\d.,]+)\s*$")
_KNOWN_TYPE_LABELS = (
    "PIX RECEBIDO", "PIX ENVIADO", "PIX QR CODE",
    "COMPRA ELO DEBITO", "PAGTO ELETRON", "TED", "DOC", "TRANSFERENCIA",
)


class BradescoPdfExtratoParser:
    bank_id   = "bradesco"
    format_id = "pdf_extrato"

    def sniff(self, data: bytes, filename: str) -> bool:
        if not filename.lower().endswith(".pdf"):
            return False
        try:
            text = _extract_text(data, max_pages=1)
        except Exception:
            return False
        return "BRADESCO" in strip_accents(text).upper()

    def parse(self, data: bytes) -> List[NormalizedRow]:
        return parse_text(_extract_text(data))


def _is_noise(ln: str) -> bool:
    upper = strip_accents(ln).upper()
    if ln.startswith(_NOISE_PREFIXES):
        return True
    if upper.startswith("DATA HISTORICO"):
        return True
    if _TOTAL_RE.match(ln):
        return True
    return False


def _is_known_type_label(ln: str) -> bool:
    upper = strip_accents(ln).upper()
    return any(upper.startswith(t) for t in _KNOWN_TYPE_LABELS)


def parse_text(text: str) -> List[NormalizedRow]:
    """Núcleo do parser, separado de _extract_text() pra poder ser testado
    sem um PDF de verdade (recebendo diretamente o texto já extraído)."""
    raw_lines = [ln.strip() for ln in text.splitlines() if ln.strip()]
    lines = [ln for ln in raw_lines if not _is_noise(ln)]

    rows: List[NormalizedRow] = []
    current_date: Optional[datetime] = None
    tipo_buffer: List[str] = []

    i = 0
    n = len(lines)
    while i < n:
        ln = lines[i]

        cod_m = _COD_LANC_RE.match(ln)
        if cod_m:
            if cod_m.group(1):
                current_date = _parse_date(cod_m.group(1))
            tipo_buffer = []
            i += 1
            continue

        amt_m = _AMOUNT_LINE_RE.match(ln)
        if amt_m:
            if amt_m.group(1):
                current_date = _parse_date(amt_m.group(1))
            tipo = clean_description(" ".join(tipo_buffer))
            tipo_buffer = []
            i += 1

            detail = ""
            if i < n:
                nxt = lines[i]
                if not _AMOUNT_LINE_RE.match(nxt) and not _COD_LANC_RE.match(nxt) \
                        and not _is_known_type_label(nxt):
                    detail = nxt
                    i += 1

            if tipo and current_date is not None:
                amount = parse_brl_amount(amt_m.group(3))
                direction = _infer_direction(tipo, detail)
                desc = clean_description(detail) or tipo
                is_inv = looks_like_investment(f"{tipo} {detail}")
                rows.append(NormalizedRow(
                    date=current_date,
                    description=desc,
                    amount=abs(amount),
                    direction=direction,
                    suggested_category="Investimentos" if is_inv else guess_category(desc),
                    suggested_payment_method=guess_payment_method(f"{tipo} {detail}"),
                    is_investment_like=is_inv,
                    raw=f"{tipo} | {ln} | {detail}".strip(" |"),
                ))
            continue

        tipo_buffer.append(ln)
        i += 1

    return rows


def _infer_direction(tipo: str, detail: str) -> str:
    detail_upper = strip_accents(detail).upper()
    if detail_upper.startswith("REM:"):
        return "entrada"
    if detail_upper.startswith("DES:"):
        return "saida"
    tipo_upper = strip_accents(tipo).upper()
    if "RECEBIDO" in tipo_upper:
        return "entrada"
    return "saida"


def _parse_date(date_str: str):
    return datetime.strptime(date_str, "%d/%m/%Y").date()


def _extract_text(data: bytes, max_pages: int = None) -> str:
    import pdfplumber
    parts = []
    with pdfplumber.open(io.BytesIO(data)) as pdf:
        pages = pdf.pages[:max_pages] if max_pages else pdf.pages
        for page in pages:
            parts.append(page.extract_text() or "")
    return "\n".join(parts)
